### ISSUE TITLE
Can’t search: MDB_READERS_FULL

### ISSUE THREAD

**Issue by irevoire** [meilisearch/meilisearch#2648]
**Describe the bug**
Hello, I was running a bunch of requests in parallel and then meilisearch throwed a few hundred times this error:
```json
{
  "code": "internal",
  "link": "https://docs.meilisearch.com/errors#internal",
  "message": "An internal error has occurred. `MDB_READERS_FULL: Environment maxreaders limit reached`.",
  "type": "internal"
}
```

I would like to know how many connection we can do at the same time?
And also this limit should be specified in the [known limits](https://docs.meilisearch.com/learn/advanced/known_limitations.html) documentation page.

**To Reproduce**
1. Run meilisearch
2. Send requests

**Expected behavior**
It execute the search

**Screenshots**

**Meilisearch version:** main branch

**Additional context**
Additional information that may be relevant to the issue.
I’m running macOS

**Comment by curquiza** [comment]
Thanks @irevoire 

Is it something we also had in the previous versions? (v0.27.0 or v0.28.0)

**Comment by Kerollmops** [comment]
Hey,

I had an idea that maybe we could send a [429 Too Many Requests](http://web.archive.org/web/20210812043115/https://httpstatuses.com/429) instead of an internal error. But we could find more and better ideas by looking [at the Retry-After guide by Mozilla](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After).

**Comment by irevoire** [comment]
Maybe we could create a « search service » that holds **ONE** read transaction and make every search goes through.
And every 2s, it could drop it and open a new one to get the latest version of the DB?

**Comment by Kerollmops** [comment]
Why not but I find it a little bit more complex as there is more synchronization to do between thread and workers, also doing so forces us to share the read transaction between thread (`Sync`) and enable the related cargo feature of heed.

**Comment by irevoire** [comment]
Yeah, that's sure, but it does fix the issue instead of letting the user handle it.

> share the read transaction between thread 

Do you think it would solve the issue to have one read transaction _by thread_ instead of one for everyone?
But I'm afraid some people will have more than 126 threads and thus will get the error right away 😬

**Comment by Kerollmops** [comment]
What I worry about is that it is not recommended to keep a read transaction for too long as [it can block write transactions](https://pijul.org/posts/2021-01-15-sanakirja-locks/) (as there can only be 1 read transaction and 1 write transaction at the same time) and keeping a read transaction for too long can force the database to grow in size.

Therefore, I am worried about how will you make sure to drop the read transaction after 2 seconds if there is no read transaction and only a writing one. Will you use channels to send read transaction requests to a function that is running in a separate thread that is dynamically handling the 2 seconds timeout to drop the transaction? That's a quite complex system.

**Comment by irevoire** [comment]
Huum, I don't think we'll need such a complex system, but I'm not sure yet; I think before choosing any solution, it would be good to make a POC of this one since it's the only one really solving the issue instead of making it appear later.

**Comment by Kerollmops** [comment]
Another solution could simply be to either:
 - Let the user specify how many parallel readers he wants to allow using a CLI argument.
 - Or multiply the number of available CPUs by the current default value of 126 (allowing 126 readers **by** thread).

**Comment by irevoire** [comment]
But why is there a limit in the beginning? Will it throw an error if we ask for too many readers?

**Comment by Kerollmops** [comment]
So after reading a little bit more about [the consequences of setting this parameter to a high value](https://web.archive.org/web/20200112155415/https://twitter.com/armon/status/534867803426533376), we can as it doesn't impact performances.

**Comment by curquiza** [comment]
Closing this in favor of https://github.com/meilisearch/meilisearch/issues/2786

**Comment by davelarkan** [comment]
Hi @Kerollmops and @curquiza 👋 
I was just wondering if I could get some clarification on this one.
What exactly do we mean by "at the same time"?

I'm trying to understand what the implication is for the cloud team.
Is this, "126 requests started within the same millisecond"? Or do we mean that 126 is the maximum number of requests that can be in a "processing state" by Meilisearch concurrently?

**Comment by curquiza** [comment]
> Or do we mean that 126 is the maximum number of requests that can be in a "processing state" by Meilisearch concurrently?

This one, this is the number of readers, so the number of "queued" request that Meilisearch can have

#2786 is closed so now it's not 126 but 1024 😇 (released in v0.30.0)
